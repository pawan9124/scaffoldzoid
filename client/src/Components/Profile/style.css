.profile_section {
  display: flex;
  flex-direction: column;
  margin: 15px;
}
.profile_name_section {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
}
.profile_about {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
}
.profile_image_section {
  display: flex;
  flex-direction: column;
  align-items: center;
  flex: 1;
}
.profile_image_section > .upload_icon {
  top: -40px;
  position: relative;
  height: 35px;
  width: 35px;
  cursor: pointer;
}
.profile_image {
  height: 150px !important;
  width: 150px !important;
}
.profile_description_section {
  flex: 2;
}
.profile_chart {
}
.profile_button {
  width: 100px;
  margin: 15px;
  cursor: pointer;
}
