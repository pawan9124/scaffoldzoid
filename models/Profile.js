import mongoose from "mongoose";

const Profile = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "users",
    required: true,
  },
  avatar: {
    type: String,
  },
  description: {
    type: String,
    required: true,
  },
});

export default mongoose.model("profiles", Profile);
