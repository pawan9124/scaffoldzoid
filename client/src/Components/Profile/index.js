import React, { useEffect, useState } from "react";
import Avatar from "@material-ui/core/Avatar";
import TextField from "@material-ui/core/TextField";
import Button from "../Reusable/Button";
import { withRouter } from "react-router-dom";
import Chart from "../Chart";
import validator from "validator";
import ImageUploader from "../Reusable/ImageUploader";
import { useSelector, useDispatch } from "react-redux";
import { useHistory } from "react-router-dom";
import { saveProfile, fetchProfileById } from "../../actions/profileActions";
import "./style.css";

function Profile(props) {
  const [sentImageFiles, setSentImageFiles] = useState([]);
  const [imageFiles, setImageFiles] = useState([]);
  const [description, setDescription] = useState("");
  const [errors, setErrors] = useState({});

  /* Get the user details from the redux store for seller usage*/
  const userDetails = useSelector((state) => {
    return state.auth.user;
  });
  /* Get the current profile details from the redux store  for both seller and customer (get name of seller)*/
  const currentProfile = useSelector((state) => state?.profile?.profiles[0]);
  const dispatchProps = useDispatch();
  const history = useHistory();

  /* use effect to fetch the user profile information in redux store*/
  useEffect(() => {
    dispatchProps(fetchProfileById(props.match.params.id));
  }, [props.match.params.id]);

  /* Use effect to set the current profile to the fields (image,description) */
  useEffect(() => {
    if (currentProfile && Object.keys(currentProfile).length > 0) {
      setImageFiles([{ imagePreviewUrl: currentProfile.avatar }]);
      setDescription(currentProfile.description);
    }
  }, [currentProfile]);

  /* Handle the submit button to save the profile and image */
  const handleSubmitButton = (e) => {
    const tempErrors = {};
    e.preventDefault();
    if (validator.isEmpty(description)) {
      tempErrors.description = "Please enter description";
    }
    /* Sending data in the form  data with the image */
    if (Object.keys(tempErrors).length === 0) {
      const fd = new FormData();
      fd.append("description", description);
      fd.append("user", userDetails.id);
      if (sentImageFiles[0] && sentImageFiles[0][0]) {
        fd.append("avatar", sentImageFiles[0][0]);
      }
      dispatchProps(saveProfile(fd, history));
    } else {
      setErrors(tempErrors);
    }
  };
  return (
    <div className="profile_section" data-test="ProfileComponent">
      {userDetails?.isSeller && <h5>Hello, {userDetails?.username}</h5>}
      {!userDetails?.isSeller && (
        <h4>Mr/Ms. {currentProfile?.user?.username}</h4>
      )}
      {/* Profile about section */}
      {userDetails?.isSeller && props.match.params.id === userDetails?.id && (
        <div className="profile_name_section">
          <Button
            type="submit"
            onClick={handleSubmitButton}
            buttonClass="default__button profile_button"
            buttonLabel="Save"
          />
        </div>
      )}
      <div className="profile_about">
        <div className="profile_image_section">
          <Avatar
            alt="C"
            src={imageFiles[0]?.imagePreviewUrl}
            className="profile_image"
          />
          {/* Seller to upload image */}
          {userDetails?.isSeller &&
            props.match.params.id === userDetails?.id && (
              <ImageUploader
                setSentImageFiles={setSentImageFiles}
                setImageFiles={setImageFiles}
              />
            )}
        </div>
        <div className="profile_description_section">
          <TextField
            id="outlined-basic"
            label={description ? "" : "Description"}
            variant="outlined"
            multiline
            rows={5}
            fullWidth={true}
            defaultValue={description}
            InputProps={{
              readOnly:
                !userDetails?.isSeller ||
                props.match.params.id !== userDetails?.id,
            }}
            onChange={(e) => setDescription(e.target.value)}
          />
          {errors.description && (
            <small className="register__validationErrors">
              {errors.description}
            </small>
          )}
        </div>
      </div>
      {/* Display chart area */}
      <div className="profile_chart">
        <h2>Chart Details</h2>
        <Chart />
      </div>
    </div>
  );
}

export default withRouter(Profile);
