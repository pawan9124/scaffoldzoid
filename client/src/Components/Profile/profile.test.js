import React from "react";
import { mount } from "enzyme";
import { findByTestAttr, checkProps, createMockStore } from "../../testUtils";
import Profile from "./index";
import renderer from "react-test-renderer";
import { withRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";

const props = {
  history: {},
  location: {},
  match: {},
  staticContext: undefined,
};
const store = createMockStore({
  auth: { isAuthenticated: true, user: {} },
  errors: {},
  profile: { profiles: {}, allProfiles: [] },
  rates: { currentRate: {}, allRates: [] },
});

describe("Taking snap shot", () => {
  it("Taking the snapshot of the Component", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <Profile {...props} />
          </Router>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe("Checking the render of component", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(
      <Provider store={store}>
        <Router>
          <Profile {...props} />
        </Router>
      </Provider>
    );
  });
  it("Should Render a component", () => {
    const component = findByTestAttr(wrapper, "ProfileComponent");
    expect(component.length).toBe(1);
  });
});

describe("Checking PropTypes", () => {
  it("Should Not throw a warning", () => {
    const expectedProps = props;

    const propsError = checkProps(Profile, expectedProps);
    expect(propsError).toBeUndefined();
  });
});
