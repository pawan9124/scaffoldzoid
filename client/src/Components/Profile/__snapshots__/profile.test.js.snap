// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Taking snap shot Taking the snapshot of the Component 1`] = `
<div
  className="profile_section"
  data-test="ProfileComponent"
>
  <h4>
    Mr/Ms. 
  </h4>
  <div
    className="profile_about"
  >
    <div
      className="profile_image_section"
    >
      <div
        className="MuiAvatar-root MuiAvatar-circle profile_image MuiAvatar-colorDefault"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiAvatar-fallback"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
          />
        </svg>
      </div>
    </div>
    <div
      className="profile_description_section"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="outlined-basic"
          id="outlined-basic-label"
        >
          Description
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
          onClick={[Function]}
        >
          <textarea
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
            defaultValue=""
            disabled={false}
            id="outlined-basic"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            readOnly={true}
            required={false}
            rows={5}
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-3"
            >
              <span>
                Description
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
  <div
    className="profile_chart"
  >
    <h2>
      Chart Details
    </h2>
    <div
      className="chart_section"
      data-test="ChartComponent"
    >
      <div
        className="chart__section_heading"
      >
        <div>
          <h3>
            Hello, 
            ! 
          </h3>
          <p>
            Manage your charts by adding, editing the existing ones and remove the types of oranges that are listed for the customers
             
          </p>
        </div>
        <p>
          List of every type of best quality oranges available from this seller, order your favorite one.
        </p>
      </div>
      <div
        className="chart_button"
      >
        <button
          className="btn  default__button profile_button btn-adjust"
          data-test="buttonComponent"
          onClick={[Function]}
        >
          Add Type
          <span
            className="btn-icon"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        className="chat_grid"
      >
        <div
          data-test="RateGridComponent"
        >
          <div
            aria-colcount={2}
            aria-rowcount={1}
            className="rdg r104f42s700-canary36"
            onScroll={[Function]}
            role="grid"
            style={
              Object {
                "--filter-row-height": "45px",
                "--header-row-height": "35px",
                "--row-height": "35px",
                "--row-width": "160px",
                "--template-columns": "80px 80px ",
              }
            }
          >
            <div
              aria-rowindex={1}
              className="rdg-header-row h1fquj5h700-canary36 h19m9p7h700-canary36"
              role="row"
            >
              <div
                aria-colindex={1}
                className="rdg-cell c1wupbe700-canary36"
                role="columnheader"
                style={
                  Object {
                    "gridColumnStart": 1,
                  }
                }
              >
                Orange Type
              </div>
              <div
                aria-colindex={2}
                className="rdg-cell c1wupbe700-canary36"
                role="columnheader"
                style={
                  Object {
                    "gridColumnStart": 2,
                  }
                }
              >
                Orange Rate (Rs/Kg)
              </div>
            </div>
            <div
              className="rdg-focus-sink f7ly7s700-canary36"
              onFocus={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
            />
            <div
              style={
                Object {
                  "height": 0,
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
